kernel/generic/src/mm/backend_elf.c

```diff
@@ -163,72 +163,65 @@ void elf_share(as_area_t *area)
 	 * Find the node in which to start linear search.
 	 */
 	if (area->flags & AS_AREA_WRITE) {
-		node = list_get_instance(list_first(&area->used_space.leaf_list),
-		    btree_node_t, leaf_link);
+		start = used_space_first(&area->used_space);
 	} else {
-		(void) btree_search(&area->used_space, start_anon, &leaf);
-		node = btree_leaf_node_left_neighbour(&area->used_space, leaf);
-		if (!node)
-			node = leaf;
+		/* Find first interval containing addresses >= start_anon */
+		start = used_space_find_gteq(&area->used_space, start_anon);
 	}
 
 	/*
 	 * Copy used anonymous portions of the area to sh_info's page map.
 	 */
 	mutex_lock(&area->sh_info->lock);
-	for (cur = &node->leaf_link; cur != &area->used_space.leaf_list.head;
-	    cur = cur->next) {
+	cur = start;
+	while (cur != NULL) {
+		uintptr_t base = cur->page;
+		size_t count = cur->count;
 		unsigned int i;
 
-		node = list_get_instance(cur, btree_node_t, leaf_link);
+		/*
+		 * Skip read-only areas of used space that are backed
+		 * by the ELF image.
+		 */
+		if (!(area->flags & AS_AREA_WRITE))
+			if (base >= entry->p_vaddr &&
+			    base + P2SZ(count) <= start_anon)
+				continue;
 
-		for (i = 0; i < node->keys; i++) {
-			uintptr_t base = node->key[i];
-			size_t count = (size_t) node->value[i];
-			unsigned int j;
+		for (i = 0; i < count; i++) {
+			pte_t pte;
+			bool found;
 
 			/*
-			 * Skip read-only areas of used space that are backed
-			 * by the ELF image.
+			 * Skip read-only pages that are backed by the
+			 * ELF image.
 			 */
 			if (!(area->flags & AS_AREA_WRITE))
 				if (base >= entry->p_vaddr &&
-				    base + P2SZ(count) <= start_anon)
+				    base + P2SZ(i + 1) <= start_anon)
```

Comment:
Could this possibly overflow?

Reply:
I don't know but I am not changing the logic in this piece of code. Just indentation and j -> i. The diff is mismatched.